rs/execution_environment/src/execution_environment.rs

```diff
@@ -3174,36 +3218,21 @@ impl ExecutionEnvironment {
     /// Aborts paused execution in the given state.
     pub fn abort_canister(&self, canister: &mut CanisterState, log: &ReplicaLogger) {
         if !canister.system_state.task_queue.is_empty() {
-            let task_queue = std::mem::take(&mut canister.system_state.task_queue);
-            canister.system_state.task_queue = task_queue
-                .into_iter()
-                .map(|task| match task {
-                    ExecutionTask::AbortedExecution { .. }
-                    | ExecutionTask::AbortedInstallCode { .. }
-                    | ExecutionTask::Heartbeat
-                    | ExecutionTask::GlobalTimer
-                    | ExecutionTask::OnLowWasmMemory => task,
-                    ExecutionTask::PausedExecution { id, .. } => {
-                        let paused = self.take_paused_execution(id).unwrap();
-                        let (input, prepaid_execution_cycles) = paused.abort(log);
-                        self.metrics.executions_aborted.inc();
-                        ExecutionTask::AbortedExecution {
-                            input,
-                            prepaid_execution_cycles,
-                        }
-                    }
-                    ExecutionTask::PausedInstallCode(id) => {
-                        let paused = self.take_paused_install_code(id).unwrap();
-                        let (message, call_id, prepaid_execution_cycles) = paused.abort(log);
-                        self.metrics.executions_aborted.inc();
-                        ExecutionTask::AbortedInstallCode {
-                            message,
-                            call_id,
-                            prepaid_execution_cycles,
-                        }
-                    }
-                })
-                .collect();
+            let paused_tasks = canister.system_state.task_queue.get_all_paused_tasks();
+
+            self.metrics
+                .executions_aborted
+                .inc_by(paused_tasks.len().try_into().unwrap());
+
+            // TODO: EXC-1730 if `PausedExecutionRegistry` becomes local
+            // we can abort all exectuions on it without `paused_tasks`.
+            let aborted_tasks = self.abort_paused_executions_and_return_tasks(paused_tasks, log);
+
+            canister
+                .system_state
+                .task_queue
+                .replace_paused_with_aborted_tasks(aborted_tasks);
```

Comment:
I know that the part of:

taking all paused tasks from TaskQueue
then aborting them in PausedExecutionRegistry
and finally replacing paused to aborted task on the task queue
Is not ideal but at the moment it is the best option in my knowledge if we want to make the task queue private. It is planned to refactor [PausedExecutionRegistry](https://dfinity.atlassian.net/browse/EXC-1730) to become local since there is no reason to be global, and at that moment we can skip the first step in this process.

Reply:
Those are a couple of very good points, I agree with you, that should be done. How about I do it in [follow-up](https://dfinity.atlassian.net/browse/EXC-1747)?

Reply:
SGTM

Although I would personally do a very basic simplification (similar to what the methods I mentioned above) in this PR already. But that's just my preference, feel free to merge as is and address it fully in a follow-up change.